File: apps/api/schemas/playbook.py
"""
Pydantic v2 schemas for the Playbook API.

A Playbook ITEM is either:
  - kind="tactic": a hand-drawn board document (``data`` = {entities, frames}),
    opened/edited in /tactics.
  - kind="round": a saved demo round (``demoId`` + ``roundNumber``), played
    back in the real 2D replay viewer.

Items live in optional folders (``folderId``). Folders are their own resource.
Mirrors ``apps/web/types/playbook.ts``. Aliases are camelCase.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class PlaybookCreate(BaseModel):
    title: str = "Untitled tactic"
    map: str = "de_mirage"
    side: Optional[str] = None
    type: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    team_id: Optional[int] = Field(None, alias="teamId")
    folder_id: Optional[int] = Field(None, alias="folderId")
    kind: str = "tactic"  # "tactic" | "round"
    demo_id: Optional[int] = Field(None, alias="demoId")
    round_number: Optional[int] = Field(None, alias="roundNumber")
    data: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True)


class PlaybookUpdate(BaseModel):
    """All fields optional — only the provided ones are written."""

    title: Optional[str] = None
    map: Optional[str] = None
    side: Optional[str] = None
    type: Optional[str] = None
    tags: Optional[list[str]] = None
    team_id: Optional[int] = Field(None, alias="teamId")
    folder_id: Optional[int] = Field(None, alias="folderId")
    kind: Optional[str] = None
    demo_id: Optional[int] = Field(None, alias="demoId")
    round_number: Optional[int] = Field(None, alias="roundNumber")
    data: Optional[dict[str, Any]] = None

    model_config = ConfigDict(populate_by_name=True)


class PlaybookSummary(BaseModel):
    """Lightweight row for the playbook library list."""

    id: int
    title: str
    map: str
    side: Optional[str] = None
    type: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    team_id: Optional[int] = Field(None, alias="teamId")
    folder_id: Optional[int] = Field(None, alias="folderId")
    kind: str = "tactic"
    demo_id: Optional[int] = Field(None, alias="demoId")
    round_number: Optional[int] = Field(None, alias="roundNumber")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = ConfigDict(populate_by_name=True)


class PlaybookResponse(PlaybookSummary):
    """Full playbook including the authored entities + frames (tactics)."""

    data: dict[str, Any] = Field(default_factory=dict)


# --------------------------------------------------------------------------
# Folders
# --------------------------------------------------------------------------


class FolderCreate(BaseModel):
    name: str = "Untitled folder"
    team_id: Optional[int] = Field(None, alias="teamId")

    model_config = ConfigDict(populate_by_name=True)


class FolderUpdate(BaseModel):
    name: Optional[str] = None

    model_config = ConfigDict(populate_by_name=True)


class FolderResponse(BaseModel):
    id: int
    name: str
    team_id: Optional[int] = Field(None, alias="teamId")
    item_count: int = Field(0, alias="itemCount")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = ConfigDict(populate_by_name=True)

File: apps/web/types/playbook.ts
// Tactical-board / Playbook types. Mirrors apps/api/schemas/playbook.py.
//
// Coordinate convention: all positions and stroke points are NORMALIZED
// to [0..1] of the radar square (x→right, y→down). The board multiplies
// by `radarSize` to get world pixels, so a playbook is independent of the
// radar image resolution. Utility coverage radii are in CS GAME UNITS and
// are converted to pixels with the per-map `scale` (units per radar px).

export type Team = "ct" | "tt";

export type EntityKind =
  | "player"
  | "smoke"
  | "molotov"
  | "incendiary"
  | "flash"
  | "he"
  | "decoy"
  | "bomb";

export type DrawTool = "select" | "pen" | "arrow" | "rect" | "circle" | "eraser";

export interface Vec2 {
  x: number;
  y: number;
}

export interface BoardEntity {
  id: string;
  kind: EntityKind;
  team?: Team;
  /** Display label, e.g. "1".."5" for players or "A Smoke". */
  label?: string;
  /** Utilities only: gameplay coverage radius in CS game units. */
  radiusWorld?: number;
  /** Players only: facing/aim direction in degrees (0 = up). */
  rot?: number;
}

export interface BoardStroke {
  id: string;
  tool: "pen" | "arrow" | "rect" | "circle";
  color: number; // 0xRRGGBB
  /** Thickness as a fraction of radarSize (so it scales with the map). */
  width: number;
  /** normalized; pen = polyline, arrow/rect/circle = [start, end] */
  points: Vec2[];
}

export interface PlaybookFrame {
  id: string;
  name?: string;
  /** Transition time (ms) to reach THIS frame from the previous one. */
  durationMs: number;
  /** entityId → normalized position. Every entity has an entry. */
  positions: Record<string, Vec2>;
  /** entityIds hidden on this frame (e.g. a smoke not yet thrown). */
  hidden?: string[];
  strokes: BoardStroke[];
  /**
   * Recorded movement paths. ``paths[entityId]`` is the polyline (normalized
   * coords) the entity travels to reach ``positions[entityId]`` when
   * transitioning INTO this frame. Captured automatically whenever the user
   * drags an entity on the board. When absent, animations fall back to a
   * straight-line ease between this frame's position and the previous one.
   */
  paths?: Record<string, Vec2[]>;
}

export interface PlaybookData {
  schemaVersion: 1;
  entities: BoardEntity[];
  frames: PlaybookFrame[];
}

// Situation category for a tactic.
export type TacticType =
  | "execute"
  | "retake"
  | "default"
  | "eco"
  | "force"
  | "pistol"
  | "anti-eco";

// A Playbook item is either a hand-drawn board tactic or a saved demo round.
export type PlaybookKind = "tactic" | "round";

// A folder grouping Playbook items.
export interface PlaybookFolder {
  id: number;
  name: string;
  teamId: number | null;
  itemCount: number;
  createdAt: string;
  updatedAt: string;
}

// ---- API shapes ----
export interface PlaybookSummary {
  id: number;
  title: string;
  map: string;
  side: Team | null;
  type: string | null;
  tags: string[];
  teamId: number | null;
  folderId: number | null;
  kind: PlaybookKind;
  demoId: number | null;
  roundNumber: number | null;
  createdAt: string;
  updatedAt: string;
}

export interface PlaybookFull extends PlaybookSummary {
  data: PlaybookData;
}

export interface PlaybookWriteBody {
  title: string;
  map: string;
  side: Team | null;
  type: string | null;
  tags: string[];
  teamId: number | null;
  folderId?: number | null;
  kind?: PlaybookKind;
  demoId?: number | null;
  roundNumber?: number | null;
  data?: PlaybookData;
}

// A team the user belongs to (for sharing playbooks). `Team` is already
// the CT/T side alias above, so the entity is `TeamInfo`.
export interface TeamInfo {
  id: number;
  name: string;
  inviteCode: string;
  role: "owner" | "member";
  memberCount: number;
  createdAt: string;
}
